### PR TITLE
Refactored network communication code into a separate package.

### PR DESCRIPTION
**Pull request by tamcap**
- Minimizes code redundancies
- Detaches communication side from the app display side

Open for comments :-)

### CODE REVIEW

File: src/org/mozilla/mozstumbler/communicator/AbstractCommunicator.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    public void close() {
+        if (httpURLConnection == null) {
+            return;
+        }
+        httpURLConnection.disconnect();
+        httpURLConnection = null;
+    }
+
+    public static class HttpErrorException extends IOException {
+        private static final long serialVersionUID = -5404095858043243126L;
```
Comment (hannosch) [inline src/org/mozilla/mozstumbler/communicator/AbstractCommunicator.java:128]: What is the serialVersionUID doing here? Looks like an arbitrary number to me.

File: src/org/mozilla/mozstumbler/communicator/Searcher.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    public float getLat() {
+        if (getStatus().equals(RESPONSE_OK_TEXT)) {
+            try {
+                return Float.parseFloat(mResponse.getString(JSON_LATITUDE));
+            } catch (JSONException e) {
+                Log.e(LOGTAG, "Latitude JSON response problem: ", e);
+            }
+        }
+        return 0;
```
Comment (hannosch) [inline src/org/mozilla/mozstumbler/communicator/Searcher.java]: Should this be `0f` to be consistent with the getLon return value?

File: src/org/mozilla/mozstumbler/communicator/Searcher.java [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+                return Float.parseFloat(mResponse.getString(JSON_LONGITUDE));
+            } catch (JSONException e) {
+                Log.e(LOGTAG, "Longitude JSON response problem: ", e);
+            }
+        }
+        return 0f;
+    }
+
+    public float getAccuracy() {
+        if (getStatus()==RESPONSE_OK_TEXT) {
```
Comment (hannosch) [inline src/org/mozilla/mozstumbler/communicator/Searcher.java]: To be consistent, I'd use `getStatus().equals...` like in getLat/getLon

File: src/org/mozilla/mozstumbler/communicator/AbstractCommunicator.java (hunk 1 above)
Comment (tamcap) [inline src/org/mozilla/mozstumbler/communicator/AbstractCommunicator.java:128]: Honestly, no clue. I was keeping it in as it was in the original commit by @illarionov. I don't see any reason to keep it, really.

File: src/org/mozilla/mozstumbler/communicator/Searcher.java (hunk 2 above)
Comment (tamcap) [inline src/org/mozilla/mozstumbler/communicator/Searcher.java]: It doesn't really matter (as 0f=0 when it comes to memory representation) but I will change it for consistency sake.

File: src/org/mozilla/mozstumbler/communicator/Searcher.java (hunk 3 above)
Comment (tamcap) [inline src/org/mozilla/mozstumbler/communicator/Searcher.java]: My bad.

File: src/org/mozilla/mozstumbler/communicator/AbstractCommunicator.java (hunk 1 above)
Comment (tamcap) [inline src/org/mozilla/mozstumbler/communicator/AbstractCommunicator.java:128]: Interesting: http://stackoverflow.com/questions/7683739/why-my-exception-class-needs-to-be-serialized

Build will fail if serialVersionUID is not included.

Comment (illarionov) [inline src/org/mozilla/mozstumbler/communicator/AbstractCommunicator.java:128]: It is a random number automatically generated by IDE :) Mostly to make lint happy.

**Comment by cpeterso** [comment]
LGTM. Thanks!

File: src/org/mozilla/mozstumbler/MapActivity.java [hunk 4]
```diff
@@ -169,6 +154,7 @@ protected void onStart() {
         super.onStart();
 
         Context context = getApplicationContext();
+        mContext = context;
```
Comment (illarionov) [inline src/org/mozilla/mozstumbler/MapActivity.java:157]: Bad idea

File: src/org/mozilla/mozstumbler/sync/SyncAdapter.java [hunk 5]
```diff
@@ -130,12 +107,12 @@ private void uploadReports(boolean ignoreNetworkStatus, SyncResult syncResult) {
             return;
         }
 
-        mNickname = mPrefs.getNickname();
         Uri uri = Reports.CONTENT_URI.buildUpon()
                 .appendQueryParameter("limit", String.valueOf(REQUEST_BATCH_SIZE))
                 .build();
         queueMinId = 0;
         queueMaxId = getMaxId();
+        Submitter submitter = new Submitter(mContext);
```
Comment (illarionov) [inline src/org/mozilla/mozstumbler/sync/SyncAdapter.java:115]: getContext()?

File: src/org/mozilla/mozstumbler/sync/SyncAdapter.java [hunk 6]
```diff
@@ -145,40 +122,31 @@ private void uploadReports(boolean ignoreNetworkStatus, SyncResult syncResult) {
...
-                cursor.close();
-                if (batch != null) {
-                    queueMinId = batch.maxId;
-                }else {
-                    queueMinId += REQUEST_BATCH_SIZE;
-                }
+            }
+            submitter.close();
+
+            cursor.close();
```
Comment (illarionov) [inline src/org/mozilla/mozstumbler/sync/SyncAdapter.java:145]: It's not a good to move cursor.close() out of finally.

File: src/org/mozilla/mozstumbler/communicator/Submitter.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            result = true;
+        } catch (HttpErrorException ex) {
+            if (ex.isTemporary())
+            {
+                isTemp = true;
+            }
+        } catch (ConnectTimeoutException ex) {
+            isTemp = true;
+        } catch (IOException ex) {
+            isTemp = false;
```
Comment (illarionov) [inline src/org/mozilla/mozstumbler/communicator/Submitter.java:49]: Any IOException should be considered temporary (including ConnectTimeoutException)

**Comment by tamcap** [comment]
Let me look into that, and probably open a new PR.

**Pull request merged**